gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/event/lineage/LineageInfo.java

```
@@ -0,0 +1,200 @@
+ *   {@link org.apache.gobblin.source.extractor.Extractor}, and destination,
+ *   represented by {@link org.apache.gobblin.writer.DataWriter} or {@link org.apache.gobblin.publisher.DataPublisher}
+ * </p>
+ *
+ * <p>
+ *   The general flow is:
+ *   <ol>
+ *     <li> source sets its {@link DatasetDescriptor} to each work unit </li>
+ *     <li> destination puts its {@link DatasetDescriptor} to the work unit </li>
+ *     <li> load and send all lineage events from all states </li>
+ *     <li> purge lineage info from all states </li>
+ *   </ol>
+ * </p>
+ */
+public final class LineageInfo {
+  static final String BRANCH = "branch";
+
+  private static final Gson GSON = new Gson();
+  private static final String NAME_KEY = "name";
+
+  private LineageInfo() {
+  }
+
+  /**
+   * Set source {@link DatasetDescriptor} of a lineage event
+   *
+   * <p>
+   *   Only the {@link org.apache.gobblin.source.Source} or its {@link org.apache.gobblin.source.extractor.Extractor}
+   *   is supposed to set the source for a work unit of a dataset
+   * </p>
+   *
+   * @param state state about a {@link org.apache.gobblin.source.workunit.WorkUnit}
+   *
+   */
+  public static void setSource(DatasetDescriptor source, State state) {
+    state.setProp(getKey(NAME_KEY), source.getName());
+    state.setProp(getKey(LineageEventBuilder.SOURCE), GSON.toJson(source));
+  }
+
+  /**
+   * Put a {@link DatasetDescriptor} of a destination dataset to a state
+   *
+   * <p>
+   *   Only the {@link org.apache.gobblin.writer.DataWriter} or {@link org.apache.gobblin.publisher.DataPublisher}
+   *   is supposed to put the destination dataset information. Since different branches may concurrently put,
+   *   the method is implemented to be threadsafe
+   * </p>
+   */
+  public static void putDestination(DatasetDescriptor destination, int branchId, State state) {
+    if (!hasLineageInfo(state)) {
+      return;
```

Should we throw an exception here? In which case we will putDestination without having lineageInfo setup in the state?

---

Lineage info shouldn't fail a job. I think a better option is to add a log warn message